File: Exercise2/foo.c
#include <pthread.h>
#include <stdio.h>

int i = 0;
pthread_mutex_t mtx;
// Note the return type: void*
void* incrementingThreadFunction(){
    // TODO: increment i 1_000_000 times
    int j=0;
    for (j; j<999998; j++){
      pthread_mutex_lock(&mtx);
        i++;
      pthread_mutex_unlock(&mtx);
    }
    return NULL;
}

void* decrementingThreadFunction(){
    // TODO: decrement i 1_000_000 times
    int j=0;
    for (j; j<999999; j++){
        pthread_mutex_lock(&mtx);
        i--;
        pthread_mutex_unlock(&mtx);
    }
    return NULL;
}


int main(){
    // TODO: declare incrementingThread and decrementingThread (hint: google pthread_create)
    pthread_t incrementingThread;
    pthread_t decrementingThread;

    pthread_mutex_init (&mtx, NULL);

    pthread_create(&incrementingThread, NULL, incrementingThreadFunction, NULL);
    pthread_create(&decrementingThread, NULL, decrementingThreadFunction, NULL);

    pthread_join(incrementingThread, NULL);
    pthread_join(decrementingThread, NULL);

    pthread_mutex_destroy(&mtx);

    printf("The magic number is: %d\n", i);
    return 0;
}

File: Exercise2/foo.py
# Python 3.3.3 and 2.7.6
# python fo.py

from threading import Thread
from threading import RLock


# Potentially useful thing:
#   In Python you "import" a global variable, instead of "export"ing it when you declare it
#   (This is probably an effort to make you feel bad about typing the word "global")
i = 0
mtx = RLock()

def incrementingFunction():
    global i
    # TODO: increment i 1_000_000 times
    for x in range(0, 999998):
        mtx.acquire()
        i = i + 1
        mtx.release()

def decrementingFunction():
    global i
    # TODO: decrement i 1_000_000 times
    for x in range(0, 999999):
        mtx.acquire()
        i = i - 1
        mtx.release()

def main():
    # TODO: Something is missing here (needed to print i)
    global i
    incrementing = Thread(target = incrementingFunction, args = (),)
    decrementing = Thread(target = decrementingFunction, args = (),)

    # TODO: Start both threads
    incrementing.start()
    decrementing.start()

    incrementing.join()
    decrementing.join()

    print("The magic number is %d" % (i))


main()
